engine/executor/dmTask/dmtask.go

```diff
@@ -0,0 +1,209 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dmtask
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/dm/dm/config"
+	dmconfig "github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/engine/jobmaster/dm"
+	"github.com/pingcap/tiflow/engine/jobmaster/dm/metadata"
+	"github.com/pingcap/tiflow/engine/jobmaster/dm/runtime"
+	"github.com/pingcap/tiflow/engine/lib"
+	libModel "github.com/pingcap/tiflow/engine/lib/model"
+	"github.com/pingcap/tiflow/engine/lib/registry"
+	"github.com/pingcap/tiflow/engine/model"
+	dcontext "github.com/pingcap/tiflow/engine/pkg/context"
+	"github.com/pingcap/tiflow/engine/pkg/externalresource/broker"
+	"github.com/pingcap/tiflow/engine/pkg/p2p"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// RegisterWorker is used to register dm task to global registry
+func RegisterWorker() {
+	registry.GlobalWorkerRegistry().MustRegisterWorkerType(lib.WorkerDMDump, newTaskFactory(lib.WorkerDMDump))
+	registry.GlobalWorkerRegistry().MustRegisterWorkerType(lib.WorkerDMLoad, newTaskFactory(lib.WorkerDMLoad))
+	registry.GlobalWorkerRegistry().MustRegisterWorkerType(lib.WorkerDMSync, newTaskFactory(lib.WorkerDMSync))
+}
+
+// taskFactory create dm task
+type taskFactory struct {
+	workerType libModel.WorkerType
+}
+
+// newTaskFactory creates abstractFactory
+func newTaskFactory(workerType libModel.WorkerType) *taskFactory {
+	return &taskFactory{workerType: workerType}
+}
+
+// DeserializeConfig implements WorkerFactory.DeserializeConfig
+func (f taskFactory) DeserializeConfig(configBytes []byte) (registry.WorkerConfig, error) {
+	cfg := &dmconfig.SubTaskConfig{}
+	err := cfg.Decode(string(configBytes), true)
+	return cfg, err
+}
+
+// NewWorkerImpl implements WorkerFactory.NewWorkerImpl
+func (f taskFactory) NewWorkerImpl(ctx *dcontext.Context, workerID libModel.WorkerID, masterID libModel.MasterID, conf lib.WorkerConfig) (lib.WorkerImpl, error) {
+	baseDMTask := NewBaseDMTask(ctx, f.workerType, conf)
+	switch f.workerType {
+	case lib.WorkerDMDump:
+		return newDumpTask(baseDMTask), nil
+	case lib.WorkerDMLoad:
+		return newLoadTask(baseDMTask), nil
+	default:
+		return newSyncTask(baseDMTask), nil
+	}
+}
+
+// DMTask defines the interface for dump/load/sync
+type DMTask interface {
+	onInit(ctx context.Context) error
+	onFinished(ctx context.Context) error
+	createUnitHolder(cfg *config.SubTaskConfig) *unitHolder
+}
+
+// BaseDMTask implements some default methods for dm task
+type BaseDMTask struct {
+	DMTask
+	lib.BaseWorker
```

Comment:
we should assert struct BaseDMTask implements interface lib.BaseWorker, not embed it. If in future lib.BaseWorker implements more method we will not discover and panic on invocation.

Reply:
I just want to be consistent with cvstask. And I think master-worker is code level and job-task is a concept.